### ISSUE TITLE
the role of pie-core in the PIE family

### ISSUE THREAD

**Issue by ArneBinder** [ArneBinder/pie-core#17]
# What should be the final architecture of the PIE family?

## pie-core 
dependencies: `huggingface-hub`, `pytorch-lightning` (just for `pytorch-lightning.core.mixins.HyperparametersMixin`)

content (see #7 for most recent overview):
  - document and annotation
  - metric and statistic     
  - module (de-)serialization, i.e., hf_hub_mixin
  - (generic) auto   
  - taskmodule (including AutoTaskmodule)
  - model (including AutoModel)
  - annotation_pipeline (including AutoAnnotationPipeline)
  - common utilities (e.g. hydra and dictionary helpers)
  - common classes required by pie-modules as well as pie-datasets
    - `WithDocumentTypeMixin`: this is a bit unfortunate because it requires `pie_datasets.dataset` EDIT: https://github.com/ArneBinder/pie-datasets/pull/176 should allow to remove `convert_dataset`
    - `(Enter|Exit)Dataset(Dict)Mixin`
 
required actions: 
 - [x] increase test coverage, see #7
 - [x] create a release: https://github.com/ArneBinder/pie-core/releases/tag/v0.2.0

## pie-datasets
dependencies: `pie_core`, `datasets`
dataset_builder dependencies: `pie-modules` (for documents, annotations, and `utils.sequence_tagging`)
dev dependencies: same as dataset_builder dependencies

content:
- dataset, iterable dataset, dataset dict
- common base dataset builders (e.g., Brat)
   - with required annotations and documents
- helper methods and classes for dataset builders (e.g., Pipeline, Caster, Converter from `pie_dataset.document.processing.generic`)
   - or should this rather live in `pie-modules`?
- dataset builders (but this does not live in the actual `pie_datasets` package)

required actions:
 - [x] https://github.com/ArneBinder/pie-datasets/pull/190
 - [x] https://github.com/ArneBinder/pie-datasets/pull/192
 - [x] (BREAKING) https://github.com/ArneBinder/pie-datasets/pull/178
     - [x] (BREAKING) https://github.com/ArneBinder/pie-datasets/pull/195
 - [x] (OPTIONAL) (BREAKING) https://github.com/ArneBinder/pie-datasets/pull/145
 - [x] create breaking release
 - [x] update dataset scripts at HF hub with versions **before** the breaking release (https://github.com/ArneBinder/pie-datasets/pull/206)
 - [x] update dataset scripts at HF hub with versions **after** the breaking release (https://github.com/ArneBinder/pie-datasets/pull/207)
 - [x] https://github.com/ArneBinder/pie-datasets/issues/208
 - [x] https://github.com/ArneBinder/pie-datasets/issues/210

## ~~pie-modules~~ pie-documents
dependencies: `pie-core`
lazy dependencies (imported on usage): `nltk`, `flair` (both for sentence splitters)

content: everything **not** PyTorch-specific
 - document processing (e.g., tokenization, regex partitioner, etc.)
 - document types and annotations required for that
 - base document and annotations types
 - (document) metrics
 - (document) statistics
 
required actions:

still in pie-modules:
 - [x] integrate (document) metrics and statistics from `pytorch_ie`, see 
    - [x] https://github.com/ArneBinder/pie-modules/pull/177, 
    - [x] https://github.com/ArneBinder/pie-modules/pull/178, and 
    - [x] https://github.com/ArneBinder/pie-modules/pull/179
 - [x] create non-breaking release of pie-modules (requires https://github.com/ArneBinder/pie-modules/pull/170) with all current features as "backup"
 - [x] create non-breaking release including `pytorch-ie` using `pie-core`, i.e., https://github.com/ArneBinder/pytorch-ie/pull/456: [pie-modules v0.15.9](https://github.com/ArneBinder/pie-modules/releases/tag/v0.15.9)
 - [x] deprecate pie-modules

in pie-documents:
 - [x] use `pie_core` instead of `pytorch_ie`, see https://github.com/ArneBinder/pie-documents/pull/1
    - [x] remove models and taskmodules
    - [x] integrate annotations and documents from `pytorch_ie`
 - [x] create release, see [pie-documents v0.1.0](https://github.com/ArneBinder/pie-documents/releases/tag/v0.1.0)

## PyTorch-IE
dependencies: `torch`, `torchmetrics`, `pie-core`, `pie-modules`, maybe `pie-datasets`?

content: **torch related** model (and taskmodule) implementations
 - (pytorch) model
 - (pytorch) pipeline
 - models
 - taskmodules
 - datamodule?

required actions:
- [x] use `pie_core`, see https://github.com/ArneBinder/pytorch-ie/pull/456
- [x] create (hopefully) non-breaking release: [0.31.9](https://github.com/ArneBinder/pytorch-ie/releases/tag/v0.31.9)
- [x] https://github.com/ArneBinder/pytorch-ie/pull/475
- [x] create another (hopefully) non-breaking release, see [v0.31.11](https://github.com/ArneBinder/pytorch-ie/releases/tag/v0.31.11) 
- [x] create a breaking release that uses the new [pie-core v0.3.0](https://github.com/ArneBinder/pie-core/releases/tag/v0.3.0), see [v0.32.0](https://github.com/ArneBinder/pytorch-ie/releases/tag/v0.32.0)
- [x] (BREAKING) https://github.com/ArneBinder/pytorch-ie/pull/476
   - [x] requires pie-documents release that
       - [x] does not depend on pytorch-ie
       - [x] has annotations and documents
   - [x] remove annotations and documents and import them from `pie_modules`
   - [x] remove metrics and statistics and import them from `pie_modules`
- [x] https://github.com/ArneBinder/pytorch-ie/issues/462
- [x] https://github.com/ArneBinder/pytorch-ie/issues/459
- [x] (BREAKING) make `AutoAnnotationPipeline` work
   - [x] annotate all model implementations with `@Model.register()` (instead of `@PyTorchIEModel.register()`)
   - [x] remove `auto_model_class = AutoPyTorchIEModel` from `PyTorchIEPipeline` 
   - [x] remove `AutoPyTorchIEModel`
- [ ] ~~(OPTIONAL) increase test coverage~~
- [ ] ~~(OPTIONAL) increase mypy coverage~~
- [x] remove backwards compatibilities and create breaking release, see https://github.com/ArneBinder/pytorch-ie/releases/tag/v0.33.0

## Template repo

- [x] https://github.com/ArneBinder/pytorch-ie-hydra-template-1/pull/213

## Future

requirements (blockers) for all PIE, but non-PyTorch projects:
 - [x] `pie-modules` release implicitly including `pie-core` (https://github.com/ArneBinder/pie-modules/pull/176), i.e., including
     - [x] a `pytorch-ie` release including https://github.com/ArneBinder/pytorch-ie/pull/456 which requires
         - [x] https://github.com/ArneBinder/pie-core/issues/7
 - [x] https://github.com/ArneBinder/pie-datasets/pull/178 (requires above, i.e., "`pie-modules` release implicitly including `pie-core`") 

### SKLearn-IE
similar to PyTorch-IE, but with scikit-learn models and taskmodules instead of torch 
dependencies: `scikit-learn`, `pie_core`, `pie_modules`, maybe `pie_datasets`
content: **scikit-learn related** model (and taskmodule) implementations, i.e., Python modules
 - model with `SKLearnIEModel`
 - pipeline with `SKLearnIEPipeline`
 - models with `SKLearnIEModel` implementations
 - taskmodules with respective `TaskModule` implementations

### LLM-IE (or LangchainIE?)
similar to PyTorch-IE, but with LLM models and taskmodules instead of torch (i.e., based on `langchain`)
dependencies: `langchain`, `pie_core`, `pie_modules`
content: **LLM (or just langchain?) related** model (and taskmodule) implementations, i.e., Python modules
 - model with `LangchainIEModel`
 - pipeline with `LangchainIEPipeline`
 - models with respective `LangchainIEModel` implementations
 - taskmodules with respective `TaskModule` implementations

**Comment by ArneBinder** [comment]
This is finally realized.
